### PR TITLE
remaining bulky payloads as HDF5

### LINKED ISSUE #397: Migrate bulky coefficient/matrix properties to HDF5Dataset storage
## Purpose

Identify and migrate other properties with bulky numerical data to `HDF5Dataset` storage, following the pattern established in #364.

## Scope

- Prioritize based on impact and normalization requirements
- Document migration patterns for future reference

### PR DESCRIPTION
**Pull request by EBB2675**
## Purpose

Reduce archive JSON size and serialization overhead for large matrix-valued properties by storing their numerical payloads as `HDF5Dataset` references.

## Scope

**Included**

- Migrate the following complex-valued quantities to `HDF5Dataset`:
  - `ElectronicGreensFunction.value`
  - `ElectronicSelfEnergy.value`
  - `HybridizationFunction.value`
  - `HoppingMatrix.value`
- Verify native `complex128` HDF5 serialization and reading.
- Consolidate HDF5 round-trip testing for migrated properties and molecular-orbital coefficients.

**Out of scope**

- Migrating the following moderate-sized quantities:
  - `contraction_coefficients`
  - ECP projector `coefficient`
  - `u_matrix`

These quantities are typically small, and storing them as separate HDF5 datasets would add reference and access overhead without meaningfully reducing archive size. Existing normalization logic also accesses them directly through operations such as `len()` and array indexing.

- Parser changes. No active parser assignments to the migrated quantities were found.

## Context & Links

- Resolves #397
- Follows the `HDF5Dataset` migration pattern established in #364.

## Reviewer Notes

The migrated quantities now serialize as HDF5 references instead of inline complex arrays.

## Status

- [x] Ready for review

## Breaking Changes

- [x] The migrated quantities now serialize as HDF5 references. Downstream code assuming direct array values must dereference the HDF5 datasets.

## Dependencies / Blockers

- [x] None

## Testing / Validation

- [x] Unit tests
- [x] Manual testing: verified native `complex128` HDF5 serialization and round-trip reading.

### CODE REVIEW

**Comment by ladinesa** [comment]
Can you verify that these are not used in the parsers?

File: tests/properties/test_hdf5_matrix_properties.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+class MatrixPropertiesTestEntry(EntryData):
+    electronic_greens_function = SubSection(sub_section=ElectronicGreensFunction.m_def)
+    electronic_self_energy = SubSection(sub_section=ElectronicSelfEnergy.m_def)
+    hybridization_function = SubSection(sub_section=HybridizationFunction.m_def)
+    hopping_matrix = SubSection(sub_section=HoppingMatrix.m_def)
+    molecular_orbitals = SubSection(sub_section=MolecularOrbitals.m_def)
+
+
+@pytest.fixture
+def archive_with_matrix_properties() -> Generator[
```
Comment (ladinesa) [inline tests/properties/test_hdf5_matrix_properties.py]: you should probably separate the context as a fixture

File: tests/properties/test_hdf5_matrix_properties.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+class MatrixPropertiesTestEntry(EntryData):
+    electronic_greens_function = SubSection(sub_section=ElectronicGreensFunction.m_def)
+    electronic_self_energy = SubSection(sub_section=ElectronicSelfEnergy.m_def)
+    hybridization_function = SubSection(sub_section=HybridizationFunction.m_def)
+    hopping_matrix = SubSection(sub_section=HoppingMatrix.m_def)
+    molecular_orbitals = SubSection(sub_section=MolecularOrbitals.m_def)
+
+
+@pytest.fixture
```
Comment (ladinesa) [inline tests/properties/test_hdf5_matrix_properties.py]: scope?

**Comment by EBB2675** [comment]
> Can you verify that these are not used in the parsers?

yes, i have checked:

Wannier90 creates a `HoppingMatrix` section and parses complex hopping data, but the assignment to `HoppingMatrix.value` is commented out:
https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/blob/d0b9830f3c810253aa0e52e4e54387168f2de48b/src/nomad_simulation_parsers/schema_packages/wannier90.py#L204

**Comment by ndaelman-hu** [comment]
> > Can you verify that these are not used in the parsers?
> 
> yes, i have checked:
> 
> Wannier90 creates a `HoppingMatrix` section and parses complex hopping data, but the assignment to `HoppingMatrix.value` is commented out: https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/blob/d0b9830f3c810253aa0e52e4e54387168f2de48b/src/nomad_simulation_parsers/schema_packages/wannier90.py#L204

Would the write go through when re-enabled? Just as a test.

File: tests/properties/test_hdf5_matrix_properties.py (hunk 1 above)
Comment (EBB2675) [reply tests/properties/test_hdf5_matrix_properties.py]: done, the context is now a module-scoped fixture, while each test still gets its own archive

File: tests/properties/test_hdf5_matrix_properties.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    electronic_greens_function = SubSection(sub_section=ElectronicGreensFunction.m_def)
+    electronic_self_energy = SubSection(sub_section=ElectronicSelfEnergy.m_def)
+    hybridization_function = SubSection(sub_section=HybridizationFunction.m_def)
+    hopping_matrix = SubSection(sub_section=HoppingMatrix.m_def)
+    molecular_orbitals = SubSection(sub_section=MolecularOrbitals.m_def)
+
+
+@pytest.fixture(scope='module')
+def server_context() -> Generator[ServerContext, None, None]:
+    upload_id = f'test_upload_matrix_properties_h5_{create_uuid()}'
```
Comment (ladinesa) [inline tests/properties/test_hdf5_matrix_properties.py]: make this then generic.

Comment (ladinesa) [reply tests/properties/test_hdf5_matrix_properties.py]: oh i see this is in test matrix, not sure if you have the same test in the other files, we could perhaps reuse this fixture, put it in conftest

File: tests/properties/test_hdf5_matrix_properties.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        ),
+        'hybridization_function': np.array(
+            [[[[9.0 + 10.0j, 11.0 - 12.0j]]]], dtype=np.complex128
+        ),
+        'hopping_matrix': np.array(
+            [[[13.0 + 14.0j, 15.0 - 16.0j]]], dtype=np.complex128
+        ),
+    }
+    mo_coefficients = np.array([[1.0, 0.0], [0.1, 0.9]], dtype=np.float64)
+    mo_coefficients_im = np.array([[0.0, 0.0], [0.2, -0.2]], dtype=np.float64)
```
Comment (ndaelman-hu) [inline tests/properties/test_hdf5_matrix_properties.py]: The coverage here is kind of light. Would also test edge cases like empty data (not tested by the HDF5Dataset infra to my knowledge) or whether type mismatch (e.g. assigning int array when complex128 expected) are captured.

**Review by ndaelman-hu — APPROVED** [review]
It's probably more prudent to not move e.g. eigenvalues as we often have to touch them.
I would also move up your shape verification helpers to the utils.

While these quantities did not have any prior shapes, any future targets might:

  - ElectronicGreensFunction.value: [n_kpoints, n_frequencies, n_orbitals, n_orbitals]
  - ElectronicSelfEnergy.value: [n_kpoints, n_frequencies, n_orbitals, n_orbitals]
  - HybridizationFunction.value: [n_kpoints, n_frequencies, n_orbitals, n_orbitals]
  - HoppingMatrix.value: [n_wigner_seitz_points, n_orbitals, n_orbitals]

**Comment by EBB2675** [comment]
> > > Can you verify that these are not used in the parsers?
> > 
> > 
> > yes, i have checked:
> > Wannier90 creates a `HoppingMatrix` section and parses complex hopping data, but the assignment to `HoppingMatrix.value` is commented out: https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/blob/d0b9830f3c810253aa0e52e4e54387168f2de48b/src/nomad_simulation_parsers/schema_packages/wannier90.py#L204
> 
> Would the write go through when re-enabled? Just as a test.

yes, checked :+1:

**Comment by EBB2675** [comment]
hi again gentlemen, @ladinesa @ndaelman-hu @JFRudzinski 

i think I addressed the review comments:

- Moved the server context fixture to `tests/conftest.py` as `hdf5_server_context` with a generic upload id so other test modules can reuse it.
- Added `assert_hdf5_dataset_matches(dataset_ref, expected, dtype=None)` to `conftest.py` to avoid repeating the open/assert block.
- Added tests for: empty array, dtype passthrough, and shape checks for `HoppingMatrix` and `ElectronicGreensFunction`.
- On dtype: `HDF5Dataset` does not enforce it. Moving from `Quantity(type=np.complex128)` means that automatic check is gone, so `complex64` gets stored as `complex64`. I added a test that documents this behavior
- updated the descriptions for the four migrated quantities with the conventional dataset layouts, phrased as conventional layouts since the actual dimensions may depend on the represented spaces / `space_id`

File: tests/properties/test_hdf5_matrix_properties.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    )
+    return archive, data
+
+
+# ---------------------------------------------------------------------------
+# Tests
+# ---------------------------------------------------------------------------
+
+
+class TestMatrixHDF5Storage:
```
Comment (ladinesa) [inline tests/properties/test_hdf5_matrix_properties.py:71]: just on the naming, we normally mirror the test with the structure of the code, so I think it would be better if we keep this.
